src/backend/distributed/commands/create_distributed_table.c

```diff
@@ -212,6 +225,299 @@ create_reference_table(PG_FUNCTION_ARGS)
 
 
 /*
+ * upgrade_reference_table accepts a distributed table which has only one shard and
+ * replicates it across all nodes to create a reference table. It also modifies related
+ * metadata to mark the table as reference explicitly.
+ */
+Datum
+upgrade_reference_table(PG_FUNCTION_ARGS)
+{
+	Oid relationId = PG_GETARG_OID(0);
+	List *shardIntervalList = NIL;
+	ShardInterval *shardInterval = NULL;
+	uint64 shardId = INVALID_SHARD_ID;
+
+	if (!IsDistributedTable(relationId))
+	{
+		char *relationName = get_rel_name(relationId);
+		ereport(ERROR, (errcode(ERRCODE_INVALID_PARAMETER_VALUE),
+						errmsg("cannot create reference table"),
+						errdetail("Relation %s is not distributed", relationName),
+						errhint("Instead, you can use; "
+								"create_reference_table('%s');", relationName)));
+	}
+
+	if (PartitionMethod(relationId) != DISTRIBUTE_BY_HASH)
+	{
+		ereport(ERROR, (errcode(ERRCODE_FEATURE_NOT_SUPPORTED),
+						errmsg("cannot create reference table"),
+						errdetail("Only hash distributed tables can be upgraded to "
+								  "reference tables.")));
+	}
+
+	shardIntervalList = LoadShardIntervalList(relationId);
+	if (shardIntervalList->length != 1)
+	{
+		ereport(ERROR, (errcode(ERRCODE_FEATURE_NOT_SUPPORTED),
+						errmsg("cannot create reference table"),
+						errdetail("Only tables with one shard can be upgraded to "
+								  "reference tables.")));
+	}
+
+	shardInterval = (ShardInterval *) linitial(shardIntervalList);
+	shardId = shardInterval->shardId;
+
+	LockShardDistributionMetadata(shardId, ExclusiveLock);
+	LockShardResource(shardId, ExclusiveLock);
+
+	ReplicateReferenceTableToAllNodes(relationId);
+
+	PG_RETURN_VOID();
+}
+
+
+/*
+ * ReplicateReferenceTableToAllNodes accepts an implicit reference table and replicates
+ * it to all worker nodes. It assumes that caller of this function ensures that given
+ * table is an implicit reference table (i.e. hash distributed table with one shard).
+ */
+static void
+ReplicateReferenceTableToAllNodes(Oid relationId)
+{
+	List *shardIntervalList = LoadShardIntervalList(relationId);
+	ShardInterval *shardInterval = (ShardInterval *) linitial(shardIntervalList);
+	uint64 shardId = shardInterval->shardId;
+
+	List *workerNodeList = WorkerNodeList();
```

Comment:
We probably need to acquire lock before getting the `workerNodeList` on `pg_dist_node` to prevent concurrent node additions.

```C
	/* acquire a lock so that no one can add/remove node concurrently */
	pgDistNode = heap_open(DistNodeRelationId(), AccessShareLock);
```

Because, upgrading to a reference table might take long time depending on the data size, and it increases likelihood of adding a node concurrently.